test/osbuildtest.py

```diff
@@ -15,21 +15,102 @@ class TestCase(unittest.TestCase):
+
+    def assertTreeDiffsEqual(self, tree_diff1, tree_diff2):
+        """
+        Asserts two tree diffs for equality.
+
+        Before assertion, the two trees are sorted, therefore order of files
+        doesn't matter.
+
+        There's a special rule for asserting differences where we don't
+        know the exact before/after value. This is useful for example if
+        the content of file is dependant on current datetime. You can use this
+        feature by putting null value in difference you don't care about.
+
+        Example:
+            "/etc/shadow": {content: ["sha256-xxx", null]}
+
+            In this case the after content of /etc/shadow doesn't matter.
+            The only thing that matters is the before content and that
+            the content modification happened.
+        """
+        tree_diff1 = _sorted_tree(tree_diff1)
+        tree_diff2 = _sorted_tree(tree_diff2)
+        try:
+            self.assertEqual(tree_diff1['added_files'], tree_diff2['added_files'])
+            self.assertEqual(tree_diff1['deleted_files'], tree_diff2['deleted_files'])
+
+            if len(tree_diff1['differences']) != len(tree_diff2['differences']):
+                raise AssertionError('length of differences different')
+
+            for (file1, differences1), (file2, differences2) in \
+                    zip(tree_diff1['differences'].items(), tree_diff2['differences'].items()):
+
+                if file1 != file2:
+                    raise AssertionError("filename different:", file1, file2)
+
+                if len(differences1) != len(differences2):
+                    raise AssertionError("length of file differences different")
+
+                for (difference1_kind, difference1_values), (difference2_kind, difference2_values) in \
+                        zip(differences1.items(), differences2.items()):
+                    if difference1_kind != difference2_kind:
+                        raise AssertionError("different difference kinds:", difference1_kind, difference2_kind)
+
+                    if difference1_values[0] is not None \
+                            and difference2_values[0] is not None \
+                            and difference1_values[0] != difference2_values[0]:
+                        raise AssertionError("first", difference1_values[0], difference2_values[0])
+
+                    if difference1_values[1] is not None \
+                            and difference2_values[1] is not None \
+                            and difference1_values[1] != difference2_values[1]:
+                        raise AssertionError("second", difference1_values[1], difference2_values[1])
+
+        except AssertionError:
+
+            # copied from TestCase.assertDictEqual
+            diff = ('\n' + '\n'.join(difflib.ndiff(
+                pprint.pformat(tree_diff1).splitlines(),
+                pprint.pformat(tree_diff2).splitlines())))
+            raise AssertionError(diff) from None
```

Comment:
`raise from None` removes the error message you've put in the `AssertionErrors` above. Is this intentional?

Reply:
Fixed by introducing `TreeAssertionError` and removing the ugly try-except-reraise thing.